python/cudf/cudf/core/dataframe.py

```diff
@@ -2602,6 +2602,72 @@ def insert(self, loc, name, value, nan_as_null=None):
 
         self._data.insert(name, value, loc=loc)
 
+    def diff(self, periods=1, axis=0):
+        """
+        First discrete difference of element.
+
+        Calculates the difference of a Dataframe element compared with another
+        element in the Dataframe (default is element in previous row).
+
+        Parameters
+        ----------
+        periods : int, default 1
+            Periods to shift for calculating difference,
+            accepts negative values.
+        axis : {0 or ‘index’, 1 or ‘columns’}, default 0
+            Take difference over rows (0) or columns (1).
+            Only row-wise (0) shift is supported.
+
+        Returns
+        -------
+        DataFrame
+            First differences of the DataFrame.
+
+        Notes
+        -----
+        Diff currently only supports numeric dtype columns.
```

Comment:
What non-numeric dtype does pandas support?

Reply:
I think it's the same here for cuDF. I'm not sure if this line may fall out of maintenance in the future if our `__sub__` supports more than just numeric types in the future.

Reply:
For example, timestamps are not numeric but I think pandas lets you perform a diff that gives timedeltas. (Need to verify this.) Does this PR exclude those types? Would it be easy to add support?

Reply:
I opened #10212 about enabling broader type support for timestamps.